synapse/handlers/oidc.py

```diff
@@ -1035,6 +1139,161 @@ def _remote_id_from_userinfo(self, userinfo: UserInfo) -> str:
+                400, "OpenID Connect Back-Channel Logout is disabled for this provider"
+            )
+
+        metadata = await self.load_metadata()
+
+        # As per OIDC Back-Channel Logout 1.0 sec. 2.4:
+        #   A Logout Token MUST be signed and MAY also be encrypted. The same
+        #   keys are used to sign and encrypt Logout Tokens as are used for ID
+        #   Tokens. If the Logout Token is encrypted, it SHOULD replicate the
+        #   iss (issuer) claim in the JWT Header Parameters, as specified in
+        #   Section 5.3 of [JWT].
+        alg_values = metadata.get("id_token_signing_alg_values_supported", ["RS256"])
+        jwt = JsonWebToken(alg_values)
+
+        # As per sec. 2.6:
+        #    3. Validate the iss, aud, and iat Claims in the same way they are
+        #       validated in ID Tokens.
+        # Which means the audience should contain Synapse's client_id and the
+        # issuer should be the IdP issuer
+        claim_options = {
+            "iss": {"values": [self.issuer]},
+            "aud": {"values": [self.client_id]},
+        }
+
+        logger.debug("Attempting to decode JWT logout_token %r", logout_token)
+
+        # Try to decode the keys in cache first, then retry by forcing the keys
+        # to be reloaded
+        jwk_set = await self.load_jwks()
+        try:
+            claims = jwt.decode(
+                logout_token,
+                key=jwk_set,
+                claims_cls=LogoutToken,
+                claims_options=claim_options,
+            )
+        except ValueError:
+            logger.info("Reloading JWKS after decode error")
+            jwk_set = await self.load_jwks(force=True)  # try reloading the jwks
+            claims = jwt.decode(
+                logout_token,
+                key=jwk_set,
+                claims_cls=LogoutToken,
+                claims_options=claim_options,
+            )
+
+        logger.debug("Decoded logout_token JWT %r; validating", claims)
+        claims.validate(leeway=120)  # allows 2 min of clock skew
+
+        # As per sec. 2.6:
+        #    4. Verify that the Logout Token contains a sub Claim, a sid Claim,
+        #       or both.
+        #    5. Verify that the Logout Token contains an events Claim whose
+        #       value is JSON object containing the member name
+        #       http://schemas.openid.net/event/backchannel-logout.
+        #    6. Verify that the Logout Token does not contain a nonce Claim.
+        # This is all verified by the LogoutToken claims class, so at this
+        # point the `sid` claim exists and is a string.
+        sid: str = claims.get("sid")
+
+        # Find in the store if we have a device associated with that SID
+        devices = await self._store.get_devices_by_auth_provider_session_id(
+            auth_provider_id=self.idp_id,
+            auth_provider_session_id=sid,
+        )
+        for device in devices:
+            user_id = device["user_id"]
```

Comment:
This feels a little fishy -- should we be asserting that the user information matches, or do we not get any of that back and only receive the session ID?

Reply:
So, the logout token has two interesting claims: the `sid` that identifies the unique session and the `sub` that identifies the user.
A valid logout token *must* have at least one of them but can also have both. Keycloak sends both for example.

We could check the `sub` claim *if present*, since we're already storing it via the `user_external_ids` table.
One of the reason I did not do that right away, is that while the default is to use `sub` claim for the `external_id`, it can actually be mapped to something else in the `user_mapping_provider`, with the `subject_claim` config with the default mapper.
So either we assume that in an OIDC context no one with a sane mind would use something else than the `sub` claim as `external_id`, or we have to run the mapper first here to be sure to have the right `external_id`, and this is very tricky because [it requires a proper "UserInfo"](https://github.com/matrix-org/synapse/blob/2f053f3f82ca174cc1c858c75afffae51af8ce0d/synapse/handlers/oidc.py#L1262-L1273), which we don't have here.

Keep in mind, the only reason there is a way to use something else than the `sub` claim is for non-OIDC OAuth 2.0 providers, so in [the docs](https://matrix-org.github.io/synapse/latest/openid.html) that would be GitHub and Gitea.

So, either:
 - we leave it as is
 - we check the `sub` claim if it is there using the `user_external_ids` table, assuming people are not mapping something else than the `sub` claim, and properly document that caveat

Reply:
> we check the `sub` claim if it is there using the `user_external_ids` table, assuming people are not mapping something else than the `sub` claim, and properly document that caveat

I suspect this isn't really going to work -- I'm positive that there are people not using the `sub` claim.

Reply:
@clokep I was referring to this for the `backchannel_logout_ignore_sub`. This is where we show a warning if the `sub` doesn't match, and ignore that check if `backchannel_logout_ignore_sub` is set to true